atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/QosCassandraReadTestSuite.java

```diff
@@ -0,0 +1,287 @@
+            .shutdownStrategy(ShutdownStrategy.AGGRESSIVE_WITH_NETWORK_CLEANUP)
+            .build();
+
+    @BeforeClass
+    public static void createTransactionManagerAndWriteData() {
+        serializableTransactionManager = TransactionManagers.builder()
+                .config(getAtlasDbConfig())
+                .runtimeConfigSupplier(QosCassandraReadTestSuite::getAtlasDbRuntimeConfig)
+                .schemas(ImmutableList.of(TodoSchema.getSchema()))
+                .userAgent("qos-test")
+                .buildSerializable();
+
+        Awaitility.await()
+                .atMost(Duration.ONE_MINUTE)
+                .pollInterval(Duration.ONE_SECOND)
+                .until(serializableTransactionManager::isInitialized);
+
+        serializableTransactionManager
+                .runTaskWithRetry((transaction) -> {
+                    writeNTodosOfSize(transaction, 200, 1_000);
+                    return null;
+                });
+        serializableTransactionManager.close();
+    }
+
+    public static void writeNTodosOfSize(Transaction transaction, int numTodos, int size) {
+        Map<Cell, byte[]> write = new HashMap<>();
+        for (int i = 0; i < numTodos; i++) {
+            Cell thisCell = Cell.create(ValueType.FIXED_LONG.convertFromJava(random.nextLong()),
+                    TodoSchema.todoTextColumn());
+            write.put(thisCell, ValueType.STRING.convertFromJava(getTodoOfSize(size)));
+        }
+        transaction.put(TodoSchema.todoTable(), write);
+    }
+
+    @Before
+    public void setup() {
+        serializableTransactionManager = TransactionManagers.builder()
+                .config(getAtlasDbConfig())
+                .runtimeConfigSupplier(QosCassandraReadTestSuite::getAtlasDbRuntimeConfig)
+                .schemas(ImmutableList.of(TodoSchema.getSchema()))
+                .userAgent("qos-test")
+                .buildSerializable();
+
+        Awaitility.await()
+                .atMost(Duration.ONE_MINUTE)
+                .pollInterval(Duration.ONE_SECOND)
+                .until(serializableTransactionManager::isInitialized);
+
+        AtlasDbMetrics.setMetricRegistry(new MetricRegistry());
+    }
+
+    @Test
+    public void shouldBeAbleToReadSmallAmountOfBytesIfDoesNotExceedLimit() {
+        assertThat(readOneBatchOfSize(1)).hasSize(1);
+    }
+
+    @Test
+    public void shouldBeAbleToReadSmallAmountOfBytesSeriallyIfDoesNotExceedLimit() {
+        IntStream.range(0, 50)
+                .forEach(i -> assertThat(readOneBatchOfSize(1)).hasSize(1));
+    }
+
+    @Test
+    public void shouldBeAbleToReadLargeAmountsExceedingTheLimitFirstTime() {
+        assertThat(readOneBatchOfSize(12)).hasSize(12);
+    }
+
+    @Test
+    public void shouldBeAbleToReadLargeAmountsExceedingTheLimitSecondTimeWithSoftLimiting() {
+        assertThat(readOneBatchOfSize(12)).hasSize(12);
+        // The second read might actually be faster as the transaction/metadata
```

Comment:
as the transaction/metadata.... what? Is cached?

Looks like we don't check that any soft limiting actually happened here. Not sure if we can.

Reply:
Yup, cached.